packages/sync/src/firehose/index.ts

```diff
@@ -0,0 +1,342 @@
+  verifyProofs,
+} from '@atproto/repo'
+import { AtUri } from '@atproto/syntax'
+import { Subscription } from '@atproto/xrpc-server'
+import {
+  type Account,
+  type Commit,
+  type Identity,
+  type RepoEvent,
+  RepoOp,
+  isAccount,
+  isCommit,
+  isIdentity,
+  isValidRepoEvent,
+} from './lexicons'
+import {
+  Event,
+  CommitMeta,
+  CommitEvt,
+  AccountEvt,
+  AccountStatus,
+  IdentityEvt,
+} from '../events'
+import { CID } from 'multiformats/cid'
+import { SyncQueue } from '../queue'
+import { didAndSeqForEvt } from '../util'
+
+export type FirehoseOptions = {
+  idResolver: IdResolver
+
+  handleEvt: (evt: Event) => Promise<void>
+  onError: (err: Error) => void
+  getCursor?: () => Promise<number | undefined>
+
+  syncQueue?: SyncQueue // sync queue getCursor takes precedence over `getCursor`
+
+  service?: string
+  subscriptionReconnectDelay?: number
+
+  unauthenticatedCommits?: boolean
+  unauthenticatedHandles?: boolean
+
+  filterCollections?: string[]
+  excludeIdentity?: boolean
+  excludeAccount?: boolean
+  excludeCommit?: boolean
+}
+
+export class Firehose {
+  private sub: Subscription<RepoEvent>
+  private abortController: AbortController
+  private destoryDefer: Deferrable
+
+  constructor(public opts: FirehoseOptions) {
+    this.destoryDefer = createDeferrable()
+    this.abortController = new AbortController()
+    this.sub = new Subscription({
+      service: opts.service ?? 'wss://bsky.network',
+      method: 'com.atproto.sync.subscribeRepos',
+      signal: this.abortController.signal,
+      getParams: async () => {
+        if (this.opts.syncQueue) {
+          return { cursor: this.opts.syncQueue.cursor }
+        } else if (opts.getCursor) {
+          const cursor = await opts.getCursor()
+          return { cursor }
+        }
+        return undefined
+      },
+      validate: (value: unknown) => {
+        try {
+          return isValidRepoEvent(value)
+        } catch (err) {
+          this.opts?.onError(new FirehoseValidationError(err, value))
+        }
+      },
+    })
+  }
+
+  async start() {
+    try {
+      for await (const evt of this.sub) {
+        if (this.opts.syncQueue) {
+          const parsed = didAndSeqForEvt(evt)
+          if (!parsed) {
+            continue
+          }
+          this.opts.syncQueue.trackEvt(parsed.did, parsed.seq, async () => {
+            const parsed = await this.parseEvt(evt)
+            for (const write of parsed) {
+              try {
+                await this.opts.handleEvt(write)
+              } catch (err) {
+                this.opts.onError(new FirehoseHandlerError(err, write))
+              }
+            }
+          })
+        } else {
+          await this.processEvt(evt)
+        }
+      }
+    } catch (err) {
+      if (err && err['name'] === 'AbortError') {
+        this.destoryDefer.resolve()
+        return
+      }
+      this.opts.onError?.(new FirehoseSubscriptionError(err))
+      await wait(3000)
```

Comment:
What is this wait up to?

Reply:
ah is supposed to use the `subscriptionReconnectDelay` with a default value of `3000`